src/main/java/org/opensearch/security/hasher/BCryptPasswordHasher.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.security.hasher;
+
+import java.nio.CharBuffer;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.Arrays;
+
+import org.opensearch.OpenSearchSecurityException;
+import org.opensearch.SpecialPermission;
+
+import com.password4j.BcryptFunction;
+import com.password4j.HashingFunction;
+import com.password4j.Password;
+import com.password4j.types.Bcrypt;
+
+import static org.opensearch.core.common.Strings.isNullOrEmpty;
+
+public class BCryptPasswordHasher implements PasswordHasher {
```

Comment:
The HashingFunction itself is implemented as a Singleton in Password4J so I am not sure it will add much benefit to make this class singleton too. Outside of testing it only gets instantiated once. There might be other factors I am not considering though.

Reply:
> Outside of testing it only gets instantiated once

This might be a good reason to have it as a singleton instance.

Reply:
@DarshitChanpura 

Though in the current form it is quite simple to change it to singleton, my reluctance is due to how the changes introduced in this PR will evolve once we bring in configurability for BCrypt and we start supporting other hashing algorithms. 

To elaborate: 

* BCryptPasswordHasher (and other implementations of PasswordHasher) will be modified so it takes in parameters for the hashing algorithms. Though it is possible to implement Singleton that can handle parameters, it will add extra complexity and it won't be an elegant solution.
* A factory (or something of the kind) will be implemented to handle instantiation of a specific PasswordHasher based on the user's configuration. Combining the two, while possible, is again extra added complexity.

Considering the above & that there is no harm in instantiating the PasswordHasher multiple times I personally dont think the extra complexity in code and in testing is worth it in this case.

Reply:
okay, in a case where there are plans to make this configurable, we can leave it as is.

Reply:
sounds good, thanks a lot for the review!